File: examples/android/CMeSdkJava/app/src/main/java/me/connect/sdk/java/sample/structmessages/StructuredMessagesAdapter.java
package me.connect.sdk.java.sample.structmessages;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatButton;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import me.connect.sdk.java.message.StructuredMessageHolder.Response;
import me.connect.sdk.java.sample.R;
import me.connect.sdk.java.sample.db.entity.StructuredMessage;

public class StructuredMessagesAdapter extends RecyclerView.Adapter<StructuredMessagesAdapter.MessageViewHolder> {
    private List<StructuredMessage> data;
    private ItemClickListener itemClickListener;

    public StructuredMessagesAdapter(ItemClickListener itemClickListener) {
        this.itemClickListener = itemClickListener;
        this.data = new ArrayList<>();
    }

    @NonNull
    @Override
    public MessageViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.struct_messages_list_item, parent, false);
        return new MessageViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull MessageViewHolder holder, int position) {
        StructuredMessage message = data.get(position);
        holder.question.setText(message.questionText);
        holder.details.setText(message.questionDetail);
        holder.buttonHolder.removeAllViews();
        if (message.selectedAnswer != null) {
            holder.selectedAnswer.setVisibility(View.VISIBLE);
            holder.selectedAnswer.setText("Selected answer: " + message.selectedAnswer);
        } else {
            holder.selectedAnswer.setVisibility(View.GONE);
            for (Response response : message.answers) {
                Button btn = new AppCompatButton(holder.itemView.getContext());
                btn.setText(response.getText());
                holder.buttonHolder.addView(btn);
                btn.setOnClickListener(v -> {
                    for (int i = 0; i < holder.buttonHolder.getChildCount(); i++) {
                        View view = holder.buttonHolder.getChildAt(i);
                        view.setEnabled(false);
                    }
                    itemClickListener.onAnswerClick(message.id, response.getText());
                });
            }
        }
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public void setData(List<StructuredMessage> data) {
        this.data = data;
        notifyDataSetChanged();
    }

    static class MessageViewHolder extends RecyclerView.ViewHolder {
        public TextView question;
        public TextView details;
        public LinearLayout buttonHolder;
        public TextView selectedAnswer;

        public MessageViewHolder(View v) {
            super(v);
            question = v.findViewById(R.id.question);
            details = v.findViewById(R.id.details);
            buttonHolder = v.findViewById(R.id.buttonHolder);
            selectedAnswer = v.findViewById(R.id.answer);
        }
    }

    public interface ItemClickListener {
        void onAnswerClick(int entryId, String answer);
    }
}

File: Examples/android/CMeSdkJava/app_rework_kt/src/main/kotlin/me/connect/sdk/java/samplekt/structmessages/StructuredMessagesAdapter.kt
package me.connect.sdk.java.samplekt.structmessages

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.widget.AppCompatButton
import androidx.core.view.children
import androidx.recyclerview.widget.RecyclerView
import me.connect.sdk.java.samplekt.db.entity.StructuredMessage
import me.connect.sdk.java.samplekt.R


class StructuredMessagesAdapter(private val itemClickListener: ItemClickListener) : RecyclerView.Adapter<StructuredMessagesAdapter.MessageViewHolder>() {
    private var data = mutableListOf<StructuredMessage>()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MessageViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.struct_messages_list_item, parent, false)
        return MessageViewHolder(v)
    }

    override fun onBindViewHolder(holder: MessageViewHolder, position: Int) {
        val message = data[position]
        holder.question.text = message.questionText
        holder.details.text = message.questionDetail
        holder.buttonHolder.removeAllViews()
        if (message.selectedAnswer != null) {
            holder.selectedAnswer.visibility = View.VISIBLE
            holder.selectedAnswer.text = "Selected answer: ${message.selectedAnswer}"
        } else {
            holder.selectedAnswer.visibility = View.GONE
            message.answers.forEach { response ->
                val btn = AppCompatButton(holder.itemView.context)
                btn.text = response.text
                holder.buttonHolder.addView(btn)
                btn.setOnClickListener { v ->
                    holder.buttonHolder.children.forEach { it.isEnabled = false }
                    itemClickListener.onAnswerClick(message.id, response.text)
                }
            }
        }
    }

    override fun getItemCount(): Int = data.size

    fun setData(messages: List<StructuredMessage>) {
        data.clear()
        data.addAll(messages)
        notifyDataSetChanged()
    }

    class MessageViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        var question: TextView = v.findViewById(R.id.question)
        var details: TextView = v.findViewById(R.id.details)
        var buttonHolder: LinearLayout = v.findViewById(R.id.buttonHolder)
        var selectedAnswer: TextView = v.findViewById(R.id.answer)

    }

    interface ItemClickListener {
        fun onAnswerClick(entryId: Int, answer: String)
    }
}